### PR TITLE
SMAI Paper response

### PR DESCRIPTION
**Pull request by florianwechsung**
Hi,

this includes the changes requested by the reviewers for our SMAI paper. The main changes are:

- Taylor Hood solver and CG Mass Inverse for the schur complement
- the Superposition example, that is considered 'realistic' flow
- changes around the stabilisation and the `wind` to deal with linearized flow problems

### CODE REVIEW

File: alfi/driver.py [hunk 1]
```diff
@@ -113,13 +117,26 @@ def run_solver(solver, res, args):
...
+                    "nu": solver.nu.values()[0],
+                    "linear_iter": float('nan'),
+                    "nonlinear_iter": float('nan'),
+                    "time": float('nan'),
+                }
+                break
+
         if args.paraview:
-            pvdf.write(solver.visprolong(solver.z.split()[0]), time=re)
+            pvdf.write(*solver.z.split(), time=re)
```
Comment (wence-) [inline alfi/driver.py:139]: This is no longer needed because Paraview does high order I suppose.

File: alfi/solver.py [hunk 2]
```diff
@@ -37,6 +37,41 @@ def apply(self, pc, x, y):
...
+        self.nu = appctx["nu"]
+        self.gamma = appctx["gamma"]
+
+    def update(self, pc):
+        pass
+
+    def apply(self, pc, x, y):
+        self.Mksp.solve(x, y)
+        scaling = float(self.nu) + float(self.gamma)
+        y.scale(-scaling)
```
Comment (wence-) [inline alfi/solver.py:70]: OK.

File: alfi/solver.py [hunk 3]
```diff
@@ -616,7 +661,7 @@ def residual(self):
         F = (
             self.nu * inner(2*sym(grad(u)), grad(v))*dx
             + self.gamma * inner(div(u), div(v))*dx
-            + self.advect * inner(dot(grad(u), u), v)*dx
+            + self.advect * inner(dot(grad(u), self.beta), v)*dx
```
Comment (wence-) [inline alfi/solver.py:665]: This is just adapting to possibly having a wind (for fully nonlinear, beta is u).

File: alfi/solver.py [hunk 4]
```diff
@@ -629,6 +674,65 @@ def function_space(self, mesh, k):
...
+        patchlu3d = "mkl_pardiso" if self.use_mkl else "umfpack"
+        patchlu2d = "petsc"
+        opts["patch_pc_patch_sub_mat_type"] = "seqaij"
+        opts["patch_sub_pc_factor_mat_solver_type"] = patchlu3d if self.tdim > 2 else patchlu2d
+
+    def distribution_parameters(self):
+        return {"partition": True, "overlap_type": (DistributedMeshOverlapType.VERTEX, 2)}
+
+
+class TaylorHoodSolver(NavierStokesSolver):
```
Comment (wence-) [inline alfi/solver.py]: I guess this is identical to the ScottVogeliusSolver except you override the `function_space` method?

File: examples/planarlattice2d/planarlattice2d.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    """ 
+    This implements the numerical example 2 from https://arxiv.org/pdf/2007.04012.pdf 
+
+
+    u = (sin(2πx)*sin(2πy), cos(2πx)*cos(2πy))
+    p = (cos(4πx)−cos(4πy))/4
+    f = − μ∆u
+    wind = u
+
+    with
```
Comment (wence-) [inline examples/planarlattice2d/planarlattice2d.py]: Missing end of sentence?

File: examples/potentialflow2d/potentialflow2d.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    This implements the numerical example 1 from https://arxiv.org/pdf/2007.04012.pdf 
+
+
+    h = x^3 - 3*x*y*y
+    u = ∇h
+    p = −(|∇h|^2)/2 + 14/5
+    f = 0
+    wind = u
+
+    with
```
Comment (wence-) [inline examples/potentialflow2d/potentialflow2d.py]: Missing end of sentence?

File: examples/superposition2d/superposition2d.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+
+    This blends example 2:
+
+    u = (sin(2πx)*sin(2πy), cos(2πx)*cos(2πy))
+    p = (cos(4πx)−cos(4πy))/4
+    wind = u + (0, 1)
+    f = − μ∆u
+
+    with
```
Comment (wence-) [inline examples/superposition2d/superposition2d.py]: Again.

**Review by wence- — COMMENTED** [review]
Looks sane, just minor comments

File: alfi/driver.py (hunk 1 above)
Comment (pefarrell) [reply alfi/driver.py:139]: Paraview's high-order support isn't great. In particular I havent figured out how to coax paraview into drawing isosurfaces using high-order information. I'm still saving refined PVD files for degree 3 for other problems ..

File: alfi/driver.py [hunk 8]
```diff
@@ -113,13 +117,26 @@ def run_solver(solver, res, args):
...
-                    checkpoint.store(solver.z, name="up_%i" % re)
+            try:
+                if from_zero_each_time:
+                    solver.z.assign(0)
+                (z, info_dict) = solver.solve(re)
+                results[re] = info_dict
+                if args.checkpoint:
+                    with DumbCheckpoint(chkptdir + "nssolution-Re-%s" % (re), mode=FILE_UPDATE) as checkpoint:
+                        checkpoint.store(solver.z, name="up_%i" % re)
+            except:
```
Comment (pefarrell) [inline alfi/driver.py:128]: Would it be better to be explicit about which exceptions we aim to catch here?

File: alfi/solver.py [hunk 9]
```diff
@@ -37,6 +37,41 @@ def apply(self, pc, x, y):
...
+        # get function spaces
+        u = TrialFunction(V)
+        v = TestFunction(V)
+        mass = assemble(inner(u, v)*dx)
+
+        Mksp = PETSc.KSP().create(comm=pc.comm)
+        Mksp.incrementTabLevel(1, parent=pc)
+        Mksp.setOptionsPrefix(prefix + "Mp_")
+        Mksp.setOperators(mass.petscmat)
+        Mksp.setUp()
```
Comment (pefarrell) [inline alfi/solver.py:56]: Shouldn't we call `setFromOptions` _before_ `setUp`? It's possible your CG mass matrix options aren't taking effect. Or if they are, you're setting up the default preconditioner (GMRES/ILU?) and then throwing it away.

File: alfi/solver.py [hunk 10]
```diff
@@ -197,23 +235,26 @@ def visprolong(u):
...
+        self.beta = self.problem.wind(self.Z.sub(0)) or u
 
 
         F = self.residual()
+        Fnostab = F
 
         """ Stabilisation """
-        wind = split(self.z_last)[0]
+        wind = self.problem.wind(self.Z.sub(0)) or u
+        Lvwind = self.problem.wind(self.Z.sub(0)) or split(self.z_last)[0]
```
Comment (pefarrell) [inline alfi/solver.py:247]: Maybe add a comment here explaining the roles of the two winds?

File: alfi/solver.py (hunk 4 above)
Comment (pefarrell) [reply alfi/solver.py]: Yes, the `residual` method looks the same as `ScottVogeliusSolver`?

File: examples/Makefile [hunk 11]
```diff
@@ -8,11 +8,26 @@ iters2dpkp0: pre
 	$(exec) iters.py --problem bfs2d --nref-start 1 --nref-end 3 --mesh coarse06.msh --discretisation pkp0 --mh uniform --k 2 --solver-type almg --stabilisation-type supg --patch star --smoothing 6 --restriction 2>&1 | tee logs/iters2dbfs06addrestrict6.log
 	$(exec) iters.py --problem bfs2d --nref-start 1 --nref-end 3 --mesh coarse03.msh --discretisation pkp0 --mh uniform --k 2 --solver-type almg --stabilisation-type supg --patch star --smoothing 6 --restriction 2>&1 | tee logs/iters2dbfs03addrestrict6.log
 
+iters2dsvlinke:pre
```
Comment (pefarrell) [inline examples/Makefile:11]: Maybe refer to all the authors?

File: examples/iters.py [hunk 12]
```diff
@@ -56,8 +69,11 @@
         dofstr = ("%.2e" % dofs[nref]).replace("e+0", r"\times 10^")
         line = ["%i" % nref, "$%s$" % dofstr]
         for re in tableres:
-            avg_ksp_iter = float(results[nref][re]["linear_iter"]
-                                 / results[nref][re]["nonlinear_iter"])
+            try:
+                avg_ksp_iter = float(results[nref][re]["linear_iter"]
+                                     / results[nref][re]["nonlinear_iter"])
+            except:
```
Comment (pefarrell) [inline examples/iters.py]: Again, be explicit here about what exceptions we're catching.

File: examples/iters.py [hunk 13]
```diff
@@ -72,7 +88,10 @@ def rnd(i):
         dofstr = ("%.2e" % dofs[nref]).replace("e+0", r"\times 10^")
         line = ["%i" % nref, "$%s$" % dofstr]
         for re in tableres:
-            avg_ksp_iter = float(results[nref][re]["time"]*60)
+            try:
+                avg_ksp_iter = float(results[nref][re]["time"]*60)
+            except:
```
Comment (pefarrell) [inline examples/iters.py]: And here.

**Review by pefarrell — APPROVED** [review]

File: alfi/solver.py (hunk 9 above)
Comment (florianwechsung) [reply alfi/solver.py:56]: This is copied pretty much straight from [pcd.py](https://github.com/firedrakeproject/firedrake/blob/master/firedrake/preconditioners/pcd.py) in firedrake. But you're right, the [petsc documentation](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/KSP/KSPSetFromOptions.html) also seems to suggest that `KSPSetFromOptions` should be called first. @wence- , can you chime in here?

Comment (wence-) [reply alfi/solver.py:56]: Ok, we should do what petsc does/recommends

**Pull request closed without merging**